src/wallet/transaction.h

```diff
@@ -11,18 +11,104 @@
 #include <wallet/ismine.h>
 #include <threadsafety.h>
 #include <tinyformat.h>
+#include <util/overloaded.h>
 #include <util/strencodings.h>
 #include <util/string.h>
 
 #include <list>
+#include <variant>
 #include <vector>
 
-struct COutputEntry;
+//! State of transaction confirmed in a block.
+struct TxStateConfirmed {
+    uint256 confirmed_block_hash;
+    int confirmed_block_height;
+    int position_in_block;
+
+    explicit TxStateConfirmed(const uint256& block_hash, int height, int index) : confirmed_block_hash(block_hash), confirmed_block_height(height), position_in_block(index) {}
+};
+
+//! State of transaction added to mempool.
+struct TxStateInMempool {
+};
+
+//! State of rejected transaction that conflicts with a confirmed block.
+struct TxStateConflicted {
+    uint256 conflicting_block_hash;
+    int conflicting_block_height;
+
+    explicit TxStateConflicted(const uint256& block_hash, int height) : conflicting_block_hash(block_hash), conflicting_block_height(height) {}
+};
+
+//! State of transaction not confirmed or conflicting with a known block and
+//! not in the mempool. May conflict with the mempool, or with an unknown block,
+//! or be abandoned, never broadcast, or rejected from the mempool for another
+//! reason.
+struct TxStateInactive {
+    bool abandoned;
+
+    explicit TxStateInactive(bool abandoned = false) : abandoned(abandoned) {}
+};
+
+//! State of transaction loaded in an unrecognized state with unexpected hash or
+//! index values. Treated as inactive (with serialized hash and index values
+//! preserved) by default, but may enter another state if transaction is added
+//! to the mempool, or confirmed, or abandoned, or found conflicting.
+struct TxStateUnrecognized {
+    uint256 block_hash;
+    int index;
+
+    TxStateUnrecognized(const uint256& block_hash, int index) : block_hash(block_hash), index(index) {}
+};
+
+//! All possible CWalletTx states
+using TxState = std::variant<TxStateConfirmed, TxStateInMempool, TxStateConflicted, TxStateInactive, TxStateUnrecognized>;
+
+//! Subset of states transaction sync logic is implemented to handle.
+using SyncTxState = std::variant<TxStateConfirmed, TxStateInMempool, TxStateInactive>;
+
+//! Interpret TxState serialized fields as a recognized state.
+static TxState TxStateInterpretSerialized(TxStateUnrecognized data)
+{
+    if (data.block_hash == uint256::ZERO) {
+        if (data.index == 0) return TxStateInactive{};
+    } else if (data.block_hash == uint256::ONE) {
+        if (data.index == -1) return TxStateInactive{/* abandoned= */ true};
+    } else if (data.index >= 0) {
+        return TxStateConfirmed{data.block_hash, /* height= */ -1, data.index};
+    } else if (data.index == -1) {
+        return TxStateConflicted{data.block_hash, /* height= */ -1};
+    }
+    return data;
+}
+
+//! Get TxState serialized block hash. Inverse of TxStateInterpretSerialized.
+static uint256 TxStateSerializedBlockHash(const TxState& state)
```

Comment:
730df28e0082 seeing the following warnings when debug-building with debian clang 11

```rake
./wallet/transaction.h:221:59: warning: field 'tx' will be initialized after field 'm_state' [-Wreorder-ctor]
    CWalletTx(CTransactionRef tx, const TxState& state) : tx(std::move(tx)), m_state(state)
                                                          ^
./wallet/transaction.h:71:16: warning: 'static' function 'TxStateInterpretSerialized' declared in header file should be declared 'static inline' [-Wunneeded-internal-declaration]
static TxState TxStateInterpretSerialized(TxStateUnrecognized data)
               ^
./wallet/transaction.h:86:16: warning: 'static' function 'TxStateSerializedBlockHash' declared in header file should be declared 'static inline' [-Wunneeded-internal-declaration]
static uint256 TxStateSerializedBlockHash(const TxState& state)
               ^
./wallet/transaction.h:98:12: warning: 'static' function 'TxStateSerializedIndex' declared in header file should be declared 'static inline' [-Wunneeded-internal-declaration]
static int TxStateSerializedIndex(const TxState& state)
           ^
4 warnings generated.
```

Reply:
> [730df28](https://github.com/bitcoin/bitcoin/commit/730df28e00822f06c9fc11c86a74f764a2f5136c) seeing the following warnings when debug-building with debian clang 11

Thanks! Should be fixed in new push